core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -308,11 +316,15 @@ private ManifestFile filterManifest(Schema tableSchema, ManifestFile manifest) {
       PartitionSpec spec = reader.spec();
       PartitionAndMetricsEvaluator evaluator =
           new PartitionAndMetricsEvaluator(tableSchema, spec, deleteExpression);
+      boolean hasDeletedFiles = manifestDeletedPositions.containsKey(manifest.path());
+      if (hasDeletedFiles) {
+        return filterManifestWithDeletedFiles(evaluator, manifest, reader);
```

Comment:
The !canContainDeletedFiles(manifest) check already optimizes the case where there is nothing to rewrite

Reply:
I think we can make the check in `canContainDeletedFiles` a bit more efficient by checking if manifestDeletedPositions is not empty before doing any of the more expensive checks.

Reply:
https://github.com/apache/iceberg/pull/11131/files#r1765979097
Same as above, we should actually probably be updating canContainDeletedFiles so we avoid the unneccessary path based comparison when the manifestDeletedPositions is not empty.